include/splay.h

```diff
@@ -100,47 +100,132 @@ template<class V>
 void
 SplayNode<V>::walk(SPLAYWALKEE * walkee, void *state)
 {
-    if (left)
-        left->walk(walkee, state);
-
-    walkee(data, state);
-
-    if (right)
-        right->walk(walkee, state);
+    // in-order walk through tree using Morris Traversal
+    SplayNode<V> *cur = const_cast<SplayNode<V> *>(this);
+    while (cur != nullptr) {
+        if (cur->left == nullptr) {
+            // no left subtree, call walkee on cur->data and
+            // descent into right subtree
+            walkee(cur->data, state);
+            cur = cur->right;
+        } else {
+            // first descent into left subtree
+
+            // find right-most child in left tree
+            SplayNode<V> *rmc = cur->left;
+            while (rmc->right && rmc->right != cur)
+                rmc = rmc->right;
+
+            if (rmc->right == nullptr) {
+                // found the right most node in the left subtree:
+                // create a temporay thread back to cur to have
+                // a way back out of the subtree
+                rmc->right = cur;
+                // descent into left subtree
+                cur = cur->left;
+            } else {
+                // found thread in left subtree pointing back to cur:
+                // this means we moved back up a thread to cur
+                // in the last "descent into right" and completed
+                // the left subtree of cur.
+                // cut the temporary thread and call walkee on cur->data
+                rmc->right = nullptr;
+                walkee(cur->data, state);
+                // and descent into right subtree
+                cur = cur->right;
+            }
+        }
+    }
 }
 
 template<class V>
 SplayNode<V> const *
 SplayNode<V>::start() const
 {
-    if (left)
-        return left->start();
+    SplayNode<V> *cur = const_cast<SplayNode<V> *>(this);
```

Comment:
Some of the const_casts appear to be unnecessary, but we can polish this (and adjust code style) later, after the primary changes have settled.

Reply:
IIRC I had to insert this cast due to compiler warnings, because this is considered const.

Reply:
Yes, `*this` is `const` here but since this code is not modifying `*this` and is not returning a modifyable part of `*this`, there should be no need to remove that `const` (after all other problems are fixed). This all will probably go away when we adjust PR code to avoid type duplication, but you do not need to worry about it for now. We will polish const/auto/style/[spelling](https://github.com/squid-cache/squid/actions/runs/5693079812/job/15433165913?pr=1431) before merging this PR.

```suggestion
    auto cur = this;
```